tls-client/main.cpp

```diff
@@ -256,22 +261,31 @@ class HelloHTTPS {
 
 
         /* Read data out of the socket */
+        int read = 0;
         do {
-            ret = mbedtls_ssl_read(&_ssl, (unsigned char *) _buffer, sizeof(_buffer));
-        } while (ret != 0 && (ret == MBEDTLS_ERR_SSL_WANT_READ ||
-                ret == MBEDTLS_ERR_SSL_WANT_WRITE));
+            ret = mbedtls_ssl_read(&_ssl, (unsigned char *) _buffer + read, 
+                                   sizeof(_buffer) - 1);
+            if (ret > 0)
+              read += ret;
+            else if (ret != MBEDTLS_ERR_SSL_WANT_READ &&
+                ret != MBEDTLS_ERR_SSL_WANT_WRITE)
+              break;
+            /* Check each of the flags */
+            _got200 = _got200 || strstr(_buffer, HTTPS_OK_STR) != NULL;
+            _gothello = _gothello || strstr(_buffer, HTTPS_HELLO_STR) != NULL;
+        } while (!_got200 || !_gothello);
```

Comment:
Because we don't know how much to read before checking for these strings or how long to run the read loop.